rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -49,25 +49,103 @@ module CfgInput implements InputSig<Location> {
   int maxSplits() { result = 0 }
 
   /** Holds if `first` is first executed when entering `scope`. */
-  predicate scopeFirst(CfgScope scope, AstNode first) {
-    scope.(CfgImpl::ControlFlowTree).first(first)
-  }
+  predicate scopeFirst(CfgScope scope, AstNode first) { scope.scopeFirst(first) }
 
   /** Holds if `scope` is exited when `last` finishes with completion `c`. */
-  predicate scopeLast(CfgScope scope, AstNode last, Completion c) {
-    scope.(CfgImpl::ControlFlowTree).last(last, c)
-  }
+  predicate scopeLast(CfgScope scope, AstNode last, Completion c) { scope.scopeLast(last, c) }
 }
 
 module CfgImpl = Make<Location, CfgInput>;
 
 import CfgImpl
 
-class FunctionTree extends StandardPostOrderTree instanceof Function {
-  override ControlFlowTree getChildNode(int i) { i = 0 and result = super.getBody() }
+/** A trivial pattern that is always guaranteed to match. */
+predicate trivialPat(Pat p) { p instanceof WildcardPat or p instanceof IdentPat }
+
+class AsmExprTree extends LeafTree instanceof AsmExpr { }
+
+class AwaitExprTree extends StandardPostOrderTree instanceof AwaitExpr {
+  override ControlFlowTree getChildNode(int i) { i = 0 and result = super.getExpr() }
+}
+
+// NOTE: `become` is a reserved but unused keyword.
+class BecomeExprTree extends StandardPostOrderTree instanceof BecomeExpr {
+  override ControlFlowTree getChildNode(int i) { i = 0 and result = super.getExpr() }
+}
+
+class BinaryOpExprTree extends StandardPostOrderTree instanceof BinaryExpr {
+  BinaryOpExprTree() { super.getOp() != "&&" and super.getOp() != "||" }
+
+  override ControlFlowTree getChildNode(int i) {
+    i = 0 and result = super.getLhs()
+    or
+    i = 1 and result = super.getRhs()
+  }
+}
+
+class LogicalOrBinaryOpExprTree extends PreOrderTree instanceof BinaryExpr {
```

Comment:
The short-circuiting operators are pre-ordered s.t. nested operators can be handled in a way where impossible paths are eliminated.